hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -842,17 +886,177 @@ case class MatrixAggregateColsByKey(child: MatrixIR, aggIR: IR) extends MatrixIR
+          }
+        }
+
+        ir.ArrayFor(
+          ir.ArrayRange(ir.I32(0), ir.I32(nKeys), ir.I32(1)),
+          colIdx,
+          rewrite(initializeIR)) },
+      { (nAggs, sequenceIR) =>
+        val colIdx = ir.genUID()
+
+        def rewrite(x: IR): IR = {
+          x match {
+            case SeqOp(i, args, aggSig) =>
+              SeqOp(
+                ir.ApplyBinaryPrimOp(ir.Add(),
+                  ir.ApplyBinaryPrimOp(
+                    ir.Multiply(),
+                    ir.ArrayRef(ir.Ref("newColumnIndices", newColumnIndicesType), ir.Ref(colIdx, TInt32())),
+                    ir.I32(nAggs)),
+                  i),
+                args, aggSig)
+            case _ =>
+              ir.MapIR(rewrite)(x)
+          }
+        }
+
+        ir.ArrayFor(
+          ir.ArrayRange(ir.I32(0), ir.I32(oldNCols), ir.I32(1)),
+          colIdx,
+          ir.Let("sa", ir.ArrayRef(ir.Ref("colValues", oldColsType), ir.Ref(colIdx, TInt32())),
+            rewrite(sequenceIR)
+          ))
+      }
+    )
+
+    val (rTypCol, makeAnnotateCol) = ir.Compile[Long, Long, Long](
+      "AGGR", aggResultTypeCol,
+      "global", oldGlobalsType,
+      postAggIRCol
+    )
+    assert(coerce[TStruct](rTypCol) == newColValueType)
+
+    val nAggsCol = rvAggsColTemplate.length
+    val rvAggsCol = new Array[RegionValueAggregator](nAggsCol * nKeys)
+    var i = 0
+    while (i < nKeys) {
+      var j = 0
+      while (j < nAggsCol) {
+        rvAggsCol(i * nAggsCol + j) = rvAggsColTemplate(j).newInstance()
+        j += 1
+      }
+      i += 1
+    }
+
+    Region.scoped { region =>
+      val rvb = new RegionValueBuilder(region)
+
+      val globals = {
+        rvb.start(oldGlobalsType)
+        rvb.addAnnotation(oldGlobalsType, oldGlobalsBc.value)
+        rvb.end()
+      }
+
+      val cols = {
+        rvb.start(oldColsType)
+        rvb.addAnnotation(oldColsType, oldColValuesBc.value)
+        rvb.end()
+      }
+
+      val colIndices = {
+        rvb.start(newColumnIndicesType)
+        rvb.startArray(newColumnIndices.length)
+        var i = 0
+        while (i < newColumnIndices.length) {
+          rvb.addInt(newColumnIndices(i))
+          i += 1
+        }
+        rvb.endArray()
+        rvb.end()
+      }
+
+      makeInitCol(0)(region, rvAggsCol, globals, false)
+      makeSeqCol(0)(region, rvAggsCol, globals, false, cols, false, colIndices, false)
+    }
+
+    val rvb = new RegionValueBuilder()
+    val annotateF = makeAnnotateCol(0)
+
+    val mapF = (a: Annotation, i: Int) => {
+      Region.scoped { region =>
+        rvb.set(region)
+
+        val aggResults = {
+          rvb.start(aggResultTypeCol)
+          rvb.startStruct()
+          var j = 0
+          while (j < nAggsCol) {
+            rvAggsCol(i * nAggsCol + j).result(rvb)
+            j += 1
+          }
+          rvb.endStruct()
+          rvb.end()
+        }
+
+        val globals = {
+          rvb.start(oldGlobalsType)
```

Comment:
can't do this! this adds the globals once per column.

Reply:
you can bring the region.scoped outside

Reply:
we should also only add the cols once -- this does it twice (once for aggs, once for annotate)

Reply:
It's not quite twice. The first time is all of the column values for the seqOp. The second time is just the new column keys (subset of original and not all values) in an order that matches the "newColumnIndices" map.